src/main/scala/quantiles/QuantileNonSample.scala

```diff
@@ -0,0 +1,252 @@
+          if (height + 1 >= curNumOfCompactors) expand()
+          val output: Array[T] = compactors(height).compact
+          output.foreach(element => compactors(height + 1).buffer = compactors(height + 1).buffer :+ element)
+          var size = 0
+          (0 until curNumOfCompactors).foreach { height =>
+            size = size + this.compactors(height).buffer.length
+          }
+          compactorActualSize = size
+          // implement lazy here
+          break
+        }
+      }
+    }
+  }
+
+  private def compare[T](o1: T, o2: T)(implicit ord: Ordering[T]): Boolean = ord.gt(o1, o2)
+
+  /**
+    * Get the map which contains the rank of all items which is currently in the sketch.
+    * @return the map which contains the rank of all items which is currently in the sketch (RankMap)
+    */
+  def getRankMap(): mutable.Map[T,Long] = {
+    val Set = output.toMap.keySet
+    var states = scala.collection.mutable.Map[T,Long]()
+    Set.foreach{item =>
+      var curWeight = 0L
+      output.foreach(tuple => {
+        if (!compare(tuple._1,item)) {
+          curWeight = curWeight + tuple._2
+        }
+      })
+      states(item) = curWeight
+    }
+    states
+  }
+
+  /**
+    * Get CDF function of sketch items.
+    * @return CDF function
+    */
+  def getCDF(): Array[(T,Double)] = {
+    val rankMap = getRankMap()
+    val tmp = rankMap.keySet.toArray
+    quickSort(tmp)
+    val totalWeight = rankMap(tmp.last)
+    var ret = ArrayBuffer[(T,Double)]()
+    tmp.foreach{ item =>
+      ret = ret :+ (item, rankMap(item).toDouble / totalWeight.toDouble)
+    }
+    ret.toArray
+  }
+
+  /**
+    * Get the rank of query item without RankMap.
+    * @param item item to query
+    * @return the estimated rank of the query item in sketch
+    */
+  def getRank(item:T): Long = {
+    var r = 0L
+    output.foreach(tuple => {
+      if (!compare(tuple._1,item)) {
+        r = r + tuple._2
+      }
+    })
+    r
+  }
+
+  /**
+    * Get the rank of query item with RankMap.
+    * @param item item to query
+    * @param rankMap the estimated rank of the query item in sketch
+    * @return
+    */
+  def getRank(item:T, rankMap:mutable.Map[T,Long]): Long = {
+    val Set = rankMap.keySet
+    val ss = collection.immutable.SortedSet[T]() ++ Set
+    var curWeight = 0L
+    ss.foreach { SetItem =>
+      if(!compare(SetItem,item)){
+        curWeight = rankMap(SetItem)
+      }
+    }
+    curWeight
+  }
+
+
+
+  /**
+    * merge two sketches into a single one
+    *
+    * @param that another sketch
+    * @return the merged sketch
+    */
+  def merge(that: QuantileNonSample[T]) : QuantileNonSample[T] = {
+    while (this.curNumOfCompactors < that.curNumOfCompactors) {
+      this.expand()
+    }
+
+    for (i <- 0 until that.curNumOfCompactors) {
+      that.compactors(i).buffer.toArray.foreach{ item=>
+        this.compactors(i).buffer = this.compactors(i).buffer :+ item
+      }
```

Comment:
:+ is used for appending one item to an array.
++ concatenates two arrays. 
Here, it seems that ++ is what we need - this really isn't a must, just a style comment

Reply:
Yes we can concatenate two ArrayBuffers with ++ instead of using a for loop with :+, have updated the code accordingly.